packages/core/src/services/chains/ChainStreamConsumer/index.ts

```diff
@@ -0,0 +1,174 @@
+  }: {
+    controller: ReadableStreamDefaultController
+    response: ChainStepResponse<StreamType>
+    config: Config
+  }) {
+    const content = ChainStreamConsumer.parseResponseText(response)
+    enqueueChainEvent(controller, {
+      event: StreamEventTypes.Latitude,
+      data: {
+        type: ChainEventTypes.Complete,
+        config,
+        documentLogUuid: response.documentLogUuid,
+        response,
+        messages: [
+          {
+            role: MessageRole.assistant,
+            toolCalls:
+              response.streamType === 'text' ? response.toolCalls || [] : [],
+            content,
+          },
+        ],
+      },
+    })
+
+    controller.close()
+  }
+
+  static chainError({
+    controller,
+    e,
+  }: {
+    controller: ReadableStreamDefaultController
+    e: unknown
+  }) {
+    const error = parseError(e)
+    enqueueChainEvent(controller, {
+      event: StreamEventTypes.Latitude,
+      data: {
+        type: ChainEventTypes.Error,
+        error: {
+          name: error.name,
+          message: error.message,
+          stack: error.stack,
+        },
+      },
+    })
+    controller.close()
+    return error
+  }
+
+  constructor({
+    controller,
+    previousCount,
+    documentLogUuid,
+  }: {
+    controller: ReadableStreamDefaultController
+    previousCount: number
+    documentLogUuid: string
+  }) {
+    this.controller = controller
+    this.previousCount = previousCount
+    this.documentLogUuid = documentLogUuid
+    this.streamConsumer = new StreamConsumer({ controller })
+  }
+
+  start({ step }: { step: ValidatedStep }) {
+    const newMessages = step.conversation.messages.slice(this.previousCount)
+    const messageCount = this.previousCount + newMessages.length
+    enqueueChainEvent(this.controller, {
+      data: {
+        type: ChainEventTypes.Step,
+        isLastStep: step.chainCompleted,
+        config: step.conversation.config as Config,
+        messages: newMessages,
+        documentLogUuid: this.documentLogUuid,
+      },
+      event: StreamEventTypes.Latitude,
+    })
+    return { messageCount, stepStartTime: Date.now() }
+  }
+
+  async consumeAI(result: Awaited<AIReturn<StreamType>>) {
+    return this.streamConsumer.call(result)
+  }
+
+  stepCompleted(response: ChainStepResponse<StreamType>) {
+    enqueueChainEvent(this.controller, {
+      event: StreamEventTypes.Latitude,
+      data: {
+        type: ChainEventTypes.StepComplete,
+        documentLogUuid: response.documentLogUuid,
+        response: response,
+      },
+    })
+  }
+
+  chainCompleted({
+    step,
+    response,
+  }: {
+    step: ValidatedStep
+    response: ChainStepResponse<StreamType>
+  }) {
+    ChainStreamConsumer.chainCompleted({
+      controller: this.controller,
+      response,
+      config: step.conversation.config as Config,
+    })
+  }
+
+  chainError(e: unknown) {
+    return ChainStreamConsumer.chainError({
+      controller: this.controller,
+      e,
+    })
+  }
+}
```

Comment:
lots of complexity no? Not saying is wrong I'm just surprised that we moved from a single chains/run file to multiple object services, could have this been done with a bit less code?

Reply:
Before all streaming logic was spread in `runChain` also duplicated in `runEvaluation`. Imo is better to have it in one place with all stream related logic centrilized